examples/mgpcg_advanced.py

```diff
@@ -1,41 +1,56 @@
 import numpy as np
-import time
 import taichi as ti
-
-real = ti.f32
-ti.init(default_fp=real, arch=ti.x64, kernel_profiler=True)
+import time
 
 
 @ti.data_oriented
 class MGPCG:
-    def __init__(self):
+    '''
+Grid-based MGPCG solver for the possion equation:
+
+    $\\nabla^2 x = k r$
+
+Credits by `@KLozes <https://github.com/KLozes>`_.
+
+See `examples/stable_fluid.py <https://github.com/taichi-dev/taichi/blob/master/examples/stable_fluid.py>`_ for usage example.
+
+.. note::
+
+    This solver only runs on CPU and CUDA backend since it requires the
+    ``pointer`` SNode. Consider replace ``pointer`` by ``bitmasked`` on Metal,
+    by ``dense`` on OpenGL.
+    '''
+    def __init__(self, dim=2, N=512, n_mg_levels=6):
+        '''
+        :parameter dim: Dimension of the fields.
+        :parameter N: Resolution of the fields.
+        :parameter n_mg_levels: Multigrid levels.
+        '''
+
         # grid parameters
         self.use_multigrid = True
 
-        self.N = 128
-        self.N_gui = 512  # gui resolution
+        self.N = N
 
-        self.n_mg_levels = 4
+        self.n_mg_levels = n_mg_levels
         self.pre_and_post_smoothing = 2
         self.bottom_smoothing = 50
-        self.dim = 3
+        self.dim = dim
 
         self.N_ext = self.N // 2  # number of ext cells set so that that total grid size is still power of 2
         self.N_tot = 2 * self.N
 
         # setup sparse simulation data arrays
-        self.r = [ti.field(dtype=real)
+        self.r = [ti.field(dtype=float)
```

Comment:
I don't think it's a confusion at all. Python use 64bit types because they use CPU only. However it's not supported very well for OpenGL and Metal. And the 32bit types can reduce memory overhead. It's a good choice for compute graphics which doesn't need too much percsion.